### ISSUE TITLE
Axons eventually all evolve to extreeme values

### ISSUE THREAD

**Issue by PRabahy** [evolvio/evolv.io#73]
![image](https://cloud.githubusercontent.com/assets/429961/20507044/478c4fe6-b026-11e6-8a54-6421130dfde8.png)

Maybe this is the correct behavior, but it appears that eventually most of the middle values in the brain go to either 0 or 1. This just seems weird to me.

On the other hand, the output variables seem to be doing exactly what I would expect. Accel and turn have stabilized on numbers around .5, Eat stabilizes on large numbers. Fight stabilizes on large negative numbers. Birth seems sort of funny... It appears to be fairly random with clusters at -3, 10-15, and multiple hundreds.

It also appears that evolve.io has an end state. After ~1300 years, my population is trending exactly with the season. As a future enhancement it would be cool if it could detect that state, and record some sort of statistics.

Maybe we could start up an IRC channel if people are interested in discussing our findings. I just joined ircs://freenode/evolve.io if anyone else wants to drop in.

**Comment by kodomazer** [comment]
I've seen that as well... What I've been doing is running long term simulations where I leave the high temp at 0.5 and slowly drop the low temp.
I'm guessing that each hidden neuron is either becoming useless or is becoming a marker for some higher level idea.  While watching them move I see them spike in the opposite direction whenever something happens, say one of the eyes goes over water.

I'm going to try giving them an extra hidden layer and upping their turn cost and see how that changes things.

**Comment by pilger15** [comment]
I think we could just reduce the nodes in the hidden layer, since most of them seem useless and the creatures still seem to evolve. Maybe Nodes could be added and removed via evolution and have a certain cost. In longtime an optimum number could be derived from this.... 

I once read that  in most cases an extra hidden layer doesn't help compared to the additional computation costs, but it would be interesting nevertheless. 

Statistics would be useful to find out what helps and what doesn't...

**Comment by kodomazer** [comment]
The question is, what constitutes a successful creature?  In more specialized simulations we have an easy metric, how far have they traveled or how long did they live.  This is a more generalized simulation so we might just have to fall back on the usual driver of evolution, a successful creature is one that has children that has children.
Something we could do is keep track of entire lineages.  Creatures with more grandchildren are more successful.  That way if we end up putting creatures in an adverse situation where all of them end up dying then we can remake one of the more successful creatures instead of starting from scratch again.

I've been thinking about trying to do a brain revamp to give it a "brain space" but don't make it fully connected.  We can give each axon a cost to have and then plop in some number of axons for primordial creatures and eventually they'll link across.  This will give some sort of speciation, since if the brains don't match up then it's possible to have stillborns.  At some point we may have to try and reduce the amount of "improper" breeding if that kills off too many.

**Comment by pilger15** [comment]
It would be great if an evolutionary tree would be created and saved every run and furthermore, if one could "load" creatures from this tree.

**Comment by PRabahy** [comment]
I think that the most concrete definition of a successful creature (in evolve.io) that I can come up with is the one that has the most biomass as a proportional descendant of it.

If it has a direct asexual descendant with 100 energy, it should get 100 points. If it has a 2 partner sexual descendant with 100 energy it should get 50 points. If it has a asexual^2 grandchild with 100 energy, it should get 100/child's parents descendant count.

To calculate this, I believe we will need to track the entire family tree.

**Comment by PRabahy** [comment]
FYI, I just registered ircs://freenode/evolve.io. I'll try to lurk there most of the time and chat when I can.

**Comment by 10carfernandez** [comment]
I've added extra hidden layers and saw a difference in behavior over the long term. But yes, the values did seem to eventually converge to either 0 or 1, though there were some values that converged to something specific like 0.2 and 0.3. 

As far as getting rid of less useful nodes, I'm not sure if getting rid of nodes with 0 would be good since it's hard to tell if the node stays at 0 all the time for some of them, or if it suddenly flickers to another value under certain situations.

As far as end states go, I think there is more of an end state when you only have one hidden layer (e.g., the creature just turning in circles and avoiding water). One difference I noted, when adding more hidden layers, was the creatures' accelerations were more variable and I noticed that some of them appeared to stop when they reached a tile with high energy, and then kept moving after they finished sucking the energy out. 

Lastly, I'm considering doing a pull request to add ice ages and mass extinctions. I've noticed that when I manually create mass extinctions, the creatures eat much more and make more use of the energy surrounding them, which creates more competition among them and ultimately, I think, speeds up evolution.

**Comment by 10carfernandez** [comment]
I did a bit of research and found [this article](http://cs231n.github.io/neural-networks-1/#actfun), which talks about the cons of using the sigmoid function as the activation function for neural networks. It's actually a good read, IMHO, so I recommend checking it out.

In short, the sigmoid function causes the weights to become over-saturated (causing extreme values of 0 or 1 to appear) during back propagation (which I believe is what the "remembering" things that @kodomazer mentioned would consist of) and using it is less desirable if the outputs are not zero-centered. The first issue can be caused by using extreme values for the weights. 

So to address this specific issue, we might want to consider one or more of the following:
1. Consider using less extreme values for the weights
2. Normalize the inputs so that they fall between 0 and 1 
3. Use a different activation function (or no activation function at all)

For 1., it looks like the current value for the randomly generated weights falls between -1 and 1. This seems reasonable, but I can test for random values between, say, -0.5 and 0.5 or -0.1 and 0.1. 

For 2., it looks like those values are already normalized, which shouldn't contribute to any over-saturation.

For 3., I can test this by using tanh(x) as the activation function and then no activation function at all. This could also tackle the 2nd issue with sigmoid(x), which is the fact that some of the outputs are zero-centered (e.g., moving forward and backwards, turning left/right, vomiting to feed babies, mouth and body hue, etc.). 

I like @kodomazer 's idea. Another idea I've thought about is implementing a deep belief network (DBN) scheme where layers could be added at some point and creatures could have a different number of layers. Just like closely related species can't produce fertile offspring when parents have different numbers of chromosomes, two creatures with different brain sizes could not reproduce with each other either.

**Comment by kodomazer** [comment]
Looking at the article, it looks like they're talking about changing weights by back propagation.  We're using "genetic evolution" to determine the weights right now.  it takes the average of the parent's weights and then may apply a small mutation to it.



I think the oversaturation is simply because it isn't detrimental to have those extreme values. I think if we make it so it has to be more finely calibrated to detect things then we might be able to avoid oversaturation at the cost of evolution speed.  

For example, if we restrict land tiles to a certain hue range and then make water tiles blue then I would expect to see more finely tuned neuron weights because they can't simply search for 0 value, but instead have to look for a specific hue.

Something else that may be interesting to do would be to change the hue selection mechanism from directly selecting the hue to choosing how much to change the hue by.  I'm guessing it will cause it to be more zero centered.  I wonder if that would have any effects on oversaturation in the hidden layers.

**Comment by 10carfernandez** [comment]
I thought about that as well. Also, the values are rounded off. So some of those extreme values might not actually be the same at later decimal points (ie, their values might still be significantly different if the inverse of the sigmoid function were applied to them). 

I forgot to talk about the things I tested. I only tested for 3., (trying different or no activation functions). To make a long story short, it looks like we should just keep the current activation function.

What I did notice, however, is how expensive it is to check if the neuron is from the last layer in the "input()" method (especially for bigger brains) because every neuron gets checked. So I was thinking that it might be good, at some point, to move the condition up a level and have an if-else statement with two nested loops, one applying the sigmoid function, and one not so that the condition is checked WIDTH times instead of WIDTH*LENGTH times.

I like the idea of restricting tile hues because of that, and also because it might make the game look more intuitive and visually appealing.

For your last idea, I don't think it would make a difference, unless it meant having less colors to choose from, which might force the creatures to better fine-tune their selection.

**Comment by kodomazer** [comment]
For the last idea, what it's saying is instead of allowing the creature to say, "I want my mouth to be this color" it makes the creature ask itself, "what color is my mouth right now, and how do I want to change it to suit my needs."  I hope that once this is all done then it will have the effects we're hoping to see and get some nicely tuned creatures.
